crates/nu-std/CONTRIBUTING.md

```diff
@@ -0,0 +1,191 @@
+[Modules chapter of the Nushell book](https://www.nushell.sh/book/modules.html). 
+Existing code in the standard library provides pretty good examples to work from.  
+
+1. Create / update the source and test files for a custom command `foo` as follows:
+   ```shell
+   # source file 
+   /path/to/nushell/crates/nu-std/lib/foo.nu  
+   # corresponding unit test file
+   /path/to/nushell/crates/nu-std/tests/test_foo.nu
+   ```
+   Thou shalt provide unit tests to cover your changes.   
+2. Typically, the source file will implement multiple subcommands and possibly a main command as well.  
+   For an example of a custom command which does not modify environment variables (but may reference them), see:
+   ```shell
+   /path/to/nushell/crates/nu-std/lib/assert.nu
+   ```
+   Note the use of `export def` to define the subcommand.  
+
+   For an example of a custom command which *does* modify environment variables, see:
+   ```shell
+   /path/to/nushell/crates/nu-std/lib/dirs.nu
+   ```
+   Note the use of `export def-env` to define the subcommand, 
+   the use of `export-env {}` to initialize the environment variable and  `let-env` to update it. 
+
+1. A `foo` command will be exposed to the user as `std foo` (at a minimum).  
+To enable this, update file `/path/to/nushell/crates/nu-std/lib/mod.rs` and add this code:
+   ```
+   export use crates/nu-std/lib/foo.nu *    # command doesn't update environment
+   export-env {
+          use crates/nu-std/lib/bar.nu *    # command *does* update environment
+   }
+   ```
+   Note use of `use *` to hoist all subcommands into `mod.nu` and thus into the `std` namespace.
+
+2. Some commands in standard library are also hoisted into the top level, non-prefixed namespace, 
+so a `foo` command can be invoked without prefix as `foo`.  
+To do this, modify `/path/to/nushell/crates/nu-std/src/lib.rs`, 
+find the initialization of the "prelude" at line 70 or thereabouts and add `(foo, foo)`.  
+(This code may be restructured soon: if you can't find it, check with the team on Discord.)  
+Note that you will need to recompile the Nushell interpreter to test this change, 
+see [Nushell Contributing#Setup](https://github.com/nushell/nushell/blob/main/CONTRIBUTING.md#setup).
+
+More general guidelines:
+
+1. Ensure your custom command provides useful help.  
+This is done with comments before the command and within the argument declaration.  
+1. Use `error make` to report can't-proceed errors to user, not `log error`.  
+2. Use `log info` to provide a verbose progress messages that the user can optionally enable for troubleshooting. 
+e.g: 
+    ```shell
+    NU_LOG_LEVEL=INFO foo # verbose messages from command foo
+    ```
+3. Use `assert` in unit tests to check for and report failures.  
+
+### Useful Commands
+
+- Run standard library tests, logging only test failures (meaning that no output is the desired outcome.):  
+  
+  ```shell
+  cd /path/to/nushell
+  NU_LOG_LEVEL=ERROR nu "crates/nu-std/tests/run.nu"
```

Comment:
here it should be `cargo run --` and not plain `nu` :yum: 

as the library is embedded in the `nu` binary, they should be in sync together => the easiest to do that is
- to use `cargo run -- ...`
- to use the `toolkit` module in the root of the project :wink:

Reply:
Let me understand better.  Is it the case that the code for std is not loaded from the `.nu` files any longer?
How then can the developer try making changes to std.nu and friends? Does s/he have to rebuild the interpreter after making a script change?
Is there any dependency on an environment variable to find any of the std `.nu` files?
Does `$env.NU_LIB_DIRS` still function to allow you to omit the full path in a `use <module>` for modules other than `std`.  How does Nushell tell the difference?

I want to assume a std developer already has some familiarity with modules in general, so I think I need to be careful to point out where they're different so s/he's not misled by too much knowlege.

In any case, I think the point of your correction is that developer must be sure to run the instance of the interpreter that's in the current source tree (thus `cargo run`, and not run the interpreter that's somewhere in his/her $env.PATH, which might be a different version.  Is that it?

Reply:
> Let me understand better.
>
:+1: 

> Is it the case that the code for std is not loaded from the `.nu` files any longer?
>
it is not.
as i said in the thread just above, the `nu-std` code loads the modules by name just before the REPL starts or a script runs.

> How then can the developer try making changes to std.nu and friends? Does s/he have to rebuild the interpreter after making a script change?
>
yes they should rebuild => hence the `cargo run`
you could have an old version of `nu` installed which could make `nu .../tests/run.nu` behave unexpectedly.
PLUS we can not rely, in the end, on the user having the files on their filesystem, because *apriori* they won't!

> Is there any dependency on an environment variable to find any of the std `.nu` files? Does `$env.NU_LIB_DIRS` still function to allow you to omit the full path in a `use <module>` for modules other than `std`. How does Nushell tell the difference?
> 
the `nu-std` loader loads the whole library before startup and make it available to the user as `std` => you just have to call `use std` to import all the library.
> **Note**
> it's a bit like the library "being in `$env.NU_LIB_DIRS`", but it's really at the binary level => `nushell` does not look up in `$env.NU_LIB_DIRS` for `std`, it just loads it directly from the binary.

> I want to assume a std developer already has some familiarity with modules in general, so I think I need to be careful to point out where they're different so s/he's not misled by too much knowlege.
> 
absolutely!
the only "new" rule to follow here is to use the name of the modules only not their paths :relieved: 

> In any case, I think the point of your correction is that developer must be sure to run the instance of the interpreter that's in the current source tree (thus `cargo run`, and not run the interpreter that's somewhere in his/her $env.PATH, which might be a different version. Is that it?
>
that's exactly right :ok_hand: 

the worse really would be having a `nu` binary which does not feature the `std` loader / relies on path to the `std` modules...
that would very likely lead to unexpected behaviour or more probably just crash :scream: 
in the end, i did not try such a thing, but from what i wrote in the loader, that does not look right :yum: 

hope all of this is a bit clearer :relieved: